runtime/src/accounts.rs

```diff
@@ -320,36 +332,96 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+        let credit_only_locks_read = credit_only_locks.read().unwrap();
+        let mut lock_counts: HashMap<&Pubkey, MutexGuard<u64>> = HashMap::new();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k) {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            } else if let Some(credit_only_lock) = credit_only_locks_read.get(&k) {
+                let lock_count = credit_only_lock.lock_count.lock().unwrap();
+                if *lock_count > 0 {
+                    error_counters.account_in_use += 1;
+                    debug!("Account in use: {:?}", k);
+                    return Err(TransactionError::AccountInUse);
+                } else {
+                    lock_counts.insert(k, lock_count);
```

Comment:
From your comment below, I realized that since we're holding the credit-debit lock for this entire function, we also know that no other threads are running this function. I think this means we don't need to hold these `lock_counts` locks. Instead, we can go through, check that all the locks will succeed, and then later regrab the locks at the end to increment the count. Sorry about the hassle and misdirection!

Reply:
Oh duh, yes, that's true. Good deal, will revert.